Another revision SDL-0230: SPP resource management (#781)
smartdevicelink/sdl_evolution: proposals/0230-spp-resource-management-for-android.md

### Detect the case where BluetoothServerSocket fails to accept a connection from head unit. 
Even though we could detect the case, we cannot increase the number of available SPP resources, because they are used by other apps.

> All we can do in this case is notify users that SPP channel runs out of available resources, and let users close some apps that may use the BluetoothSocket. It's not practical to show SPP service records that are used by Bluetooth adapter. The proposed approach is to notify users that we're running out of resources, and gives more information if needed (though this proposal does not include giving more information part).

Reviewer: ```suggestion
All we can do in this case is notify users that SPP channel runs out of available resources, and let users close some apps that may use the BluetoothSocket. It's not practical to show SPP service records that are used by Bluetooth adapter. The proposed approach is to notify users that we're running out of resources, and gives more information if needed (though this proposal does not include the giving more information part).
```

**Fig. 2: Manage notifications setting**

And the channel ("SDL Error") should be configured like below:

![manage notification category setting](../assets/proposals/0230-spp-resource-management-for-android/manage_notification_2.png)

**Fig. 3: Manage notification category setting**

All strings used for this error UX is defined in strings.xml like below:

```java
    <string name="spp_out_of_resource">Too many apps are using Bluetooth</string>
    <string name="notification_title">SmartDeviceLink</string>
    <string name="sdl_error_notification_channel_name">SDL Error</string>
```

> The strings above needs to be reviewed by steering committee.

Reviewer: ```suggestion
The strings above need to be reviewed by SDLC steering committee.
```

1. English strings (which is default language) must be properly reviewed and maintained by steering committee.
2. Localized strings are maintained by project maintainer, and steering committee will be responsible for sign off of those localized strings.

## Potential downsides

In Fig. 3 (Manage notification category setting), if user changed the behavior to "No sound", user won't notice anything even if SPP error occurred. This seems to be a potential issue, but it is up to the user's configuration.

> On the other hand, we should be very aware of the fact that SDL proxy notifies error with sound and/or vibration when SPP error occurred.

Reviewer: ```suggestion
On the other hand, we should be very aware of the fact that SDL proxy notifies error with sound and/or vibration when SPP error occurs.
```
